overlord/servicestate/quota_control.go

```diff
@@ -68,8 +103,88 @@ func ensureSnapServicesForGroup(st *state.State, grp *quota.Group, allGrps map[s
 		ensureOpts.RequireMountedSnapdSnap = true
 	}
 
-	// TODO: do we need to restart modified services ?
-	return wrappers.EnsureSnapServices(snapSvcMap, ensureOpts, nil, progress.Null)
+	grpsToRestart := []*quota.Group{}
+	appsToRestartBySnap := map[*snap.Info][]*snap.AppInfo{}
+
+	collectModifiedUnits := func(app *snap.AppInfo, grp *quota.Group, unitType string, name, old, new string) {
+		switch unitType {
+		case "slice":
+			// this slice was modified and needs to be restarted
+			grpsToRestart = append(grpsToRestart, grp)
+
+		case "service":
+			sn := app.Snap
+			appsToRestartBySnap[sn] = append(appsToRestartBySnap[sn], app)
+
+			// TODO: what about sockets and timers? activation units just start
+			// the full unit, so as long as the full unit is restarted we should
+			// be okay?
+		}
+	}
+	if err := wrappers.EnsureSnapServices(snapSvcMap, ensureOpts, collectModifiedUnits, progress.Null); err != nil {
+		return err
+	}
+
+	// now restart the slices
+
+	// TODO: should this logic move to wrappers in wrappers.RestartGroups() ?
+	systemSysd := systemd.New(systemd.SystemMode, progress.Null)
+
+	for _, grp := range grpsToRestart {
+		// TODO: what should these timeouts for stopping/restart slices be?
+		if !ensureOpts.Preseeding {
+			if err := systemSysd.Restart(grp.SliceFileName(), 5*time.Second); err != nil {
```

Comment:
Hm I think we should make this longer, like 30 seconds maybe?

Reply:
that's the timeout used by RestartServices fwiw